lib/datadog/tracing/client_ip.rb

```diff
@@ -0,0 +1,191 @@
+      # @param [HeaderCollection, #get, nil] headers A collection with the request headers.
+      # @param [String, nil] remote_ip The remote IP the request associated with the span is sent to.
+      def self.set_client_ip_tag(span, headers, remote_ip)
+        return if configuration.disabled
+
+        ip = client_address_from_request(headers, remote_ip)
+        if !configuration.header_name && ip.nil?
+          header_names = ip_headers(headers).keys
+          span.set_tag(TAG_MULTIPLE_IP_HEADERS, header_names.join(',')) unless header_names.empty?
+          return
+        end
+
+        unless valid_ip?(ip)
+          ip = extract_ip_from_full_address(ip)
+          return unless valid_ip?(ip)
+        end
+        ip = strip_zone_specifier(ip) if valid_ipv6?(ip)
+
+        span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_CLIENT_IP, ip)
+      end
+
+      TAG_MULTIPLE_IP_HEADERS = '_dd.multiple-ip-headers'.freeze
+
+      # A header collection implementation that looks up headers in a Hash.
+      class HashHeaderCollection < HeaderCollection
+        def initialize(hash)
+          super()
+          @hash = hash.transform_keys(&:downcase)
+        end
+
+        def get(header_name)
+          @hash[header_name.downcase]
+        end
+      end
+
+      def self.ip_headers(headers)
+        return {} unless headers
+
+        {}.tap do |result|
+          DEFAULT_IP_HEADERS_NAMES.each do |name|
+            value = headers.get(name)
+            next if value.nil?
+
+            result[name] = value
+          end
+        end
+      end
+
+      def self.client_address_from_request(headers, remote_ip)
+        return headers.get(configuration.header_name) if configuration.header_name && headers
+
+        ip_values_from_headers = ip_headers(headers).values
+        case ip_values_from_headers.size
+        when 0
+          remote_ip
+        when 1
+          ip_values_from_headers.first
+        end
+      end
+
+      def self.strip_zone_specifier(ipv6)
+        if /\A(.*?)%.*/ =~ ipv6
+          return Regexp.last_match(1)
+        end
+
+        ipv6
+      end
+
+      # Extracts the IP part from a full address (`ipv4:port` or `[ipv6]:port`).
+      #
+      # @param [String] address Full address to split
+      # @returns [String] The IP part of the full address.
+      def self.extract_ip_from_full_address(address)
+        if /\A\[(.*)\]:\d+\Z/ =~ address
+          return Regexp.last_match(1)
+        end
+
+        if /\A(.*):\d+\Z/ =~ address
+          return Regexp.last_match(1)
+        end
+
+        address
+      end
+
+      def self.configuration
+        Datadog.configuration.tracing.client_ip
+      end
+
+      # Determines whether the given IP is valid.
+      #
+      # @param [String] ip The IP to validate.
+      # @returns [Boolean]
+      def self.valid_ip?(ip)
+        valid_ipv4?(ip) || valid_ipv6?(ip)
+      end
+
+      # --- Section vendored from the ipaddress gem --- #
```

Comment:
If we decide to vendor third party code, we should find a way to slip the corresponding license, at least in `LICENSE-3rd_party.csv`.

Reply:
I avoided `IPAddr` initially because it has the netmask handling which I didn't want but I guess it's solvable with some verification and it spares vendoring from `ipaddress`.
I'll give `IPAddr` a shot